libs/providers/vmware.py

```
@@ -1471,12 +1477,55 @@ def clone_vm(
+        """Reconfigure cloned NICs with uppercase manual MACs back to manual address type.
+
+        VMware generates lowercase MACs by default, so only uppercase MACs need
+        reconfiguration after cloning.
+
+        Args:
+            vm (vim.VirtualMachine): The cloned VM object to reconfigure.
+            clone_vm_name (str): Name of the cloned VM (used for logging and error messages).
+            uppercase_mac_nics (set[str]): Set of NIC labels whose MACs need uppercasing.
+        """
+        if not (vm.config and vm.config.hardware and vm.config.hardware.device):
+            raise ValueError(f"Cloned VM '{clone_vm_name}' has no hardware device configuration")
```

Ruff (0.15.18)

[warning] Avoid specifying long messages outside the exception class

(TRY003)